File: android/src/main/kotlin/com/kardenvan/sber_auth_sdk_flutter/SberAuthSdkMethodCallHandler.kt
package com.kardenvan.sber_auth_sdk_flutter

import android.content.Context
import com.kardenvan.sber_auth_sdk_flutter.login.SberIdLoginFacade
import com.kardenvan.sber_auth_sdk_flutter.login.parameters.SberIdLoginParametersFactory
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class SberAuthSdkMethodCallHandler constructor(private val context: Context) {
    fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> getPlatformVersion(result)
            "authorizeWithSberId" -> authorizeWithSberId(call, result)
            else -> handleUnknownMethodCall(result)
        }
    }

    private fun authorizeWithSberId(call: MethodCall, result: Result) {
        try {
            val callArgs = call.arguments

            if (callArgs !is Map<*, *>) throw Exception("Method call arguments are invalid")

            val loginParameters = SberIdLoginParametersFactory.fromMap(callArgs)

            val loginFacade = SberIdLoginFacade(parameters = loginParameters)
            val hasLoginBegun = loginFacade.login(context)

            if (!hasLoginBegun) {
                result.error(
                    "-99",
                    "Couldn't start SberID Login. This might be caused by absence of SBOL and browsers on the device",
                    null
                )
            } else {
                result.success(null)
            }
        } catch (e: Exception) {
            handleMethodError(result, e)
        }
    }

    private fun getPlatformVersion(result: Result) {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
    }

    private fun handleMethodError(result: Result, e: Exception) {
        result.error(
            e.message ?: "-1",
            e.localizedMessage ?: "Unknown error",
            e.stackTraceToString()
        )
    }

    private fun handleUnknownMethodCall(result: Result) {
        result.notImplemented()
    }
}

File: ios/Classes/SberAuthSdkMethodCallHandler.swift
//
//  SberAuthSdkMethodCallHandler.swift
//  sber_auth_sdk_flutter
//
//  Created by apple on 16.03.2023.
//

import Foundation

class SberAuthSdkMethodCallHandler: NSObject {
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch (call.method) {
            case "getPlatformVersion":
                getPlatformVersion(result: result)
            case "authorizeWithSberId":
                authorizeWithSberId(call: call, result: result)
            default:
                handleMethodError(
                    result: result,
                    message: "Unknown method call"
                )
        }
    }

    private func authorizeWithSberId(
        call: FlutterMethodCall,
        result: FlutterResult
    ) -> Void {
        let controller = UIApplication.shared.delegate?.window??.rootViewController as? FlutterViewController
        
        if (controller == nil) {
            return handleMethodError(
                result: result,
                message: "Context is not set for the app"
            )
        }

        do {
            let callArgs = call.arguments

            if (!(callArgs is [String: Any?])) {
                throw MyError.runtimeError("Method call arguments are invalid")
            }

            let loginParameters = try SberIdLoginParametersFactory.fromMap(
                map: callArgs as! [String: Any?]
            )

            let loginFacade = SberIdLoginFacade()
            let hasLoginBegun = loginFacade.login(
                viewController: controller!,
                parameters: loginParameters
            )

            if (!hasLoginBegun) {
                result(
                    MyError.runtimeError(
                        "Couldn't start SberID Login. This might be caused by absence of SBOL and browsers on the device"
                    )
                )
            } else {
                result(nil)
            }
        } catch MyError.runtimeError(let errorMessage) {
            handleMethodError(result: result, message: errorMessage)
        } catch {
            handleMethodError(result: result, message: error.localizedDescription)
        }
    }

    private func getPlatformVersion(result: FlutterResult) {
        result("Android ${android.os.Build.VERSION.RELEASE}")
    }

    private func handleMethodError(result: FlutterResult, message: String) {
        result(MyError.runtimeError(message))
    }
}
